src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -1,28 +1,170 @@
+     * HTTP endpoint that reveals the actual API Key value for a given apiKeyId.
+     */
+    private ApiKey getApiKey(Request req, Response res) {
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        String apiKeyId = HttpUtils.getRequiredParamFromRequest(req, "apiKeyId");
+
+        // User must be admin or have the key in order to view key details.
+        if (isUserAdmin(requestingUser) || userHasKey(requestingUser.apiUser, apiKeyId)) {
+            return new ApiKey(ApiGatewayUtils.getApiKey(apiKeyId));
+        }
+        logMessageAndHalt(req, HttpStatus.FORBIDDEN_403, "User not permitted to view API key.");
+        return null;
+    }
+
+    /**
+     * Shorthand method to determine if an API user exists and has an API key.
+     */
+    private boolean userHasKey(ApiUser user, String apiKeyId) {
+        return user != null &&
+            user.apiKeys
+                .stream()
+                .filter(ApiKey -> apiKeyId.equals(ApiKey.id))
+                .findAny()
+                .orElse(null) != null;
+    }
+
+    /**
+     * Create a new API key and assign it to the provided usage plan. If no usage plan is provided use the default
+     * usage plan instead.
+     */
+    private ApiUser createApiKeyForApiUser(Request req, Response res) {
+        ApiUser targetUser = getApiUser(req);
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        String usagePlanId = req.queryParamOrDefault("usagePlanId", DEFAULT_USAGE_PLAN_ID);
+        // If requester is not an admin user, force the usage plan ID to the default and enforce key limit. A non-admin
+        // user should not be able to create an API key for any usage plan.
+        if (!isUserAdmin(requestingUser)) {
+            usagePlanId = DEFAULT_USAGE_PLAN_ID;
+            if (targetUser.apiKeys.size() >= API_KEY_LIMIT_PER_USER) {
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "User has reached API key limit.");
+            }
+        }
+        // FIXME Should an Api user be limited to one api key per usage plan (and perhaps stage)?
+        CreateApiKeyResult apiKeyResult = ApiGatewayUtils.createApiKey(targetUser.id, usagePlanId);
+        if (apiKeyResult == null || apiKeyResult.getId() == null) {
+            logMessageAndHalt(req,
+                HttpStatus.INTERNAL_SERVER_ERROR_500,
+                String.format("Unable to get AWS api key for user (%s)", targetUser),
+                null
+            );
+            return null;
+        }
+        // Add new API key to user and persist
+        targetUser.apiKeys.add(new ApiKey(apiKeyResult));
+        Persistence.apiUsers.replace(targetUser.id, targetUser);
+        return Persistence.apiUsers.getById(targetUser.id);
+    }
+
+    /**
+     * Delete an api key from a given user's list of api keys (if present) and from AWS api gateway.
+     */
+    private ApiUser deleteApiKeyForApiUser(Request req, Response res) {
```

Comment:
Do we need to return the entire user after this operation?

Reply:
I'm deferring this one to @landonreed

Reply:
Same answer as above.